### PR TITLE
[Aptos Framework] Migrate TestCoin to Coin<TestCoin>

### PR DESCRIPTION
**Pull request by movekevin**
## Motivation

Standardize TestCoin into Coin<TestCoin>. This also lays down the foundation for eventually getting ready for Mainnet (generalizing Network coins and removing testnet-specific functionalities such as Mint/Burn capabilities of TestCoin) 

### Have you read the [Contributing Guidelines on pull requests](https://github.com/aptos-labs/aptos-core/blob/main/CONTRIBUTING.md#pull-requests)?

Yes

## Test Plan

1. Unit tests + E2E tests. 
2. Manual tests: mint test coins, CLI, SDK, example interactions.

## Screenshots

Showing the new 0x1::Coin::CoinStore<0x1::TestCoin::TestCoin> resource:
image

### CODE REVIEW

File: aptos-move/framework/aptos-framework/sources/Coin.move [hunk 1]
```diff
@@ -126,18 +126,26 @@ module AptosFramework::Coin {
...
-    ) acquires BurnCapability, CoinInfo {
+    ) acquires CoinInfo {
         let account_addr = Signer::address_of(account);
         assert!(
             exists<BurnCapability<CoinType>>(account_addr),
             Errors::not_published(ENO_BURN_CAPABILITY),
         );
-        let _cap = borrow_global<BurnCapability<CoinType>>(account_addr);
 
+        burn_with_capability(coin, BurnCapability<CoinType> { });
```
Comment (zekun000) [inline aptos-move/framework/aptos-framework/sources/Coin.move:140]: this should try to acquire the capability from signer? I think we can unify it with the assertion

Comment (movekevin) [reply aptos-move/framework/aptos-framework/sources/Coin.move]: I've changed the approach here so this comment no longer applies. PTAL again 🙏

File: api/goldens/aptos_api__tests__events_test__test_get_events.json [hunk 2]
```diff
@@ -1,10 +1 @@
-[
-  {
-    "key": "0x0400000000000000000000000000000000000000000000000000000000000000000000000a550c18",
```
Comment (zekun000) [inline api/goldens/aptos_api__tests__events_test__test_get_events.json:3]: https://github.com/aptos-labs/aptos-core/blob/33c2d745f1d398310d3ffbd3e9eb5954a534ad18/api/src/tests/events_test.rs#L8

need to change this to 0x05 too

File: api/goldens/aptos_api__tests__transactions_test__test_get_txn_execute_failed_by_invalid_script_function_arguments.json [hunk 3]
```diff
@@ -6,7 +6,7 @@
   "event_root_hash": "",
   "gas_used": "1",
   "success": false,
-  "vm_status": "Transaction Executed and Committed with Error FAILED_TO_DESERIALIZE_ARGUMENT",
+  "vm_status": "Transaction Executed and Committed with Error FUNCTION_RESOLUTION_FAILURE",
```
Comment (zekun000) [inline api/goldens/aptos_api__tests__transactions_test__test_get_txn_execute_failed_by_invalid_script_function_arguments.json:9]: this looks changed too

File: api/goldens/aptos_api__tests__transactions_test__test_get_txn_execute_failed_by_missing_script_function_arguments.json [hunk 4]
```diff
@@ -6,7 +6,7 @@
   "event_root_hash": "",
   "gas_used": "1",
   "success": false,
-  "vm_status": "Transaction Executed and Committed with Error NUMBER_OF_ARGUMENTS_MISMATCH",
+  "vm_status": "Transaction Executed and Committed with Error FUNCTION_RESOLUTION_FAILURE",
```
Comment (zekun000) [inline api/goldens/aptos_api__tests__transactions_test__test_get_txn_execute_failed_by_missing_script_function_arguments.json:9]: ditto

File: aptos-move/framework/aptos-framework/sources/Genesis.move [hunk 5]
```diff
@@ -96,8 +97,12 @@ module AptosFramework::Genesis {
...
         TransactionPublishingOption::initialize(core_resource_account, initial_script_allow_list, is_open_module);
 
-        TestCoin::initialize(core_resource_account, 1000000);
-        TestCoin::mint_internal(core_resource_account, Signer::address_of(core_resource_account), 18446744073709551615);
+        // This is testnet-specific configuration and can be skipped for mainnet.
+        // Mainnet can call Coin::initialize<MainnetCoin> directly and give mint capability to the Staking module.
+        let burn_cap = TestCoin::initialize(&core_framework_account, core_resource_account);
+
+        // Give TransactionFee BurnCapability<TestCoin> so it can burn gas.
+        TransactionFee::store_test_coin_burn_cap(&core_framework_account, burn_cap);
```
Comment (zekun000) [inline aptos-move/framework/aptos-framework/sources/Genesis.move:105]: we don't want to store anything in the 0x1 address, we'll remove the account later as well.

File: aptos-move/framework/aptos-framework/sources/TransactionFee.move [hunk 6]
```diff
@@ -1,10 +1,22 @@
...
+        burn_cap: BurnCapability<TestCoin>,
+    }
+
     /// Burn transaction fees in epilogue.
-    public(friend) fun burn_fee(fee: Coin) {
-        TestCoin::burn_gas(fee);
+    public(friend) fun burn_fee(fee: Coin<TestCoin>) acquires TestCoinCapabilities {
+        Coin::burn_with_capability<TestCoin>(
+            fee,
+            &borrow_global<TestCoinCapabilities>(@AptosFramework).burn_cap,
```
Comment (zekun000) [inline aptos-move/framework/aptos-framework/sources/TransactionFee.move:15]: i don't think we need to store a wrapper here, we can have a friend function burn in TestCoin

File: aptos-move/framework/aptos-framework/sources/TypeInfo.move [hunk 7]
```diff
@@ -21,9 +21,9 @@ module AptosFramework::TypeInfo {
 
     #[test]
     fun test() {
-        let type_info = type_of<AptosFramework::Account::Account>();
+        let type_info = type_of<TypeInfo>();
```
Comment (zekun000) [inline aptos-move/framework/aptos-framework/sources/TypeInfo.move:24]: why this change?

File: crates/aptos-faucet/src/main.rs [hunk 8]
```diff
@@ -271,30 +272,35 @@ mod tests {
...
-            match script_function {
-                ScriptFunctionCall::AccountCreateAccount {
-                    auth_key: address, ..
-                } => {
-                    let mut writer = accounts.write();
-                    let previous = writer.insert(address, AccountState::new(0));
-                    assert!(previous.is_none(), "should not create account twice");
+        if let ScriptFunction(script_function) = txn.payload() {
+            // We have to manually match mint function calls since the transaction builder
+            // cannot generate encoders/decoders generic Coin::mint<CoinType> functions.
```
Comment (zekun000) [inline crates/aptos-faucet/src/main.rs:277]: hmm, I thought the transaction builder can parse generics (ty_args)

**Review by zekun000 — COMMENTED** [review]
awesome!

File: aptos-move/framework/aptos-framework/sources/TypeInfo.move (hunk 7 above)
Comment (movekevin) [reply aptos-move/framework/aptos-framework/sources/TypeInfo.move:24]: This is because of cyclic dependency Account -> TestCoin -> Coin -> TypeInfo -> Account. It's not worth it for a test :)

File: aptos-move/framework/aptos-framework/sources/TransactionFee.move (hunk 6 above)
Comment (movekevin) [reply aptos-move/framework/aptos-framework/sources/TransactionFee.move:15]: I discussed with David yesterday and it's a bit cleaner to not have friends at all. This allows all mint/burn to go through the same standardized Coin mint/burn flows.

File: aptos-move/framework/aptos-framework/sources/Genesis.move (hunk 5 above)
Comment (movekevin) [reply aptos-move/framework/aptos-framework/sources/Genesis.move:105]: This is related to the discussion on friendship vs storing caps in system accounts. This PR is doing the latter approach, but either way we don't need to have signing control over the 0x1 address. I'm planning to remove the key rotation call for core_framework_account in a separate commit once I finish fixing all test failures.

File: crates/aptos-faucet/src/main.rs (hunk 8 above)
Comment (movekevin) [reply crates/aptos-faucet/src/main.rs:277]: I built the transaction builder and it pretty much skipped the entire Coin module because all functions there are type generic unfortunately.

File: api/goldens/aptos_api__tests__events_test__test_get_events.json (hunk 2 above)
Comment (movekevin) [reply api/goldens/aptos_api__tests__events_test__test_get_events.json:3]: Nice catch thanks!

File: api/goldens/aptos_api__tests__transactions_test__test_get_txn_execute_failed_by_invalid_script_function_arguments.json (hunk 3 above)
Comment (movekevin) [reply api/goldens/aptos_api__tests__transactions_test__test_get_txn_execute_failed_by_invalid_script_function_arguments.json]: Nice catch thanks! I missed a couple of TestCoin references.

File: api/goldens/aptos_api__tests__transactions_test__test_get_txn_execute_failed_by_missing_script_function_arguments.json (hunk 4 above)
Comment (movekevin) [reply api/goldens/aptos_api__tests__transactions_test__test_get_txn_execute_failed_by_missing_script_function_arguments.json]: Nice catch thanks! I missed a couple of TestCoin references. Fixed.

File: aptos-move/e2e-testsuite/src/tests/verify_txn.rs [hunk 9]
```diff
@@ -1089,3 +1092,17 @@ fn test_type_tag_transitive_dependency_fails_verification() {
         _ => panic!("Kept transaction with an invariant violation!"),
     }
 }
+
+pub fn encode_test_coin_transfer(to: AccountAddress, amount: u64) -> TransactionPayload {
```
Comment (davidiw) [inline aptos-move/e2e-testsuite/src/tests/verify_txn.rs:1096]: we should add these to a manually written aptos_stdlib ... say aptos_stdlib_extensions

Comment (movekevin) [reply aptos-move/e2e-testsuite/src/tests/verify_txn.rs]: Updated

File: api/goldens/aptos_api__tests__events_test__test_get_events.json (hunk 2 above)
Comment (davidiw) [reply api/goldens/aptos_api__tests__events_test__test_get_events.json:3]: wtf? haha

File: api/src/tests/transactions_test.rs [hunk 10]
```diff
@@ -874,9 +890,14 @@ async fn test_get_txn_execute_failed_by_invalid_script_function_name() {
...
         "0x1",
-        "TestCoin",
+        "Coin",
         "transfer_invalid",
-        vec![],
+        vec![TypeTag::Struct(StructTag {
+            address: AccountAddress::ONE,
+            module: ident_str!("TestCoin").to_owned(),
+            name: ident_str!("TestCoin").to_owned(),
+            type_params: vec![],
```
Comment (davidiw) [inline api/src/tests/transactions_test.rs:899]: it might be good to make this into something in types along the lines of `types/src/utility_coin.rs`

File: aptos-move/framework/aptos-framework/sources/Genesis.move (hunk 5 above)
Comment (davidiw) [reply aptos-move/framework/aptos-framework/sources/Genesis.move:105]: I think it is generally safer to assume that there will be some sort of governance hosted at 0x1 or some equally impossible to generate address.

File: aptos-move/framework/aptos-framework/sources/TestCoin.move [hunk 11]
```diff
@@ -47,44 +30,34 @@ module AptosFramework::TestCoin {
...
 
-    struct SentEvent has drop, store {
-        amount: u64,
-        to: address,
-    }
+        let (mint_cap, burn_cap) = Coin::initialize_with_capabilities<TestCoin>(
+            core_framework,
+            ASCII::string(b"Test Coin"),
+            ASCII::string(b"TC"),
+            18, /* decimals */
```
Comment (davidiw) [inline aptos-move/framework/aptos-framework/sources/TestCoin.move:40]: if we do this many, we're going to think through u64 vs u128 and the perf implications of that being our default coin amount. I don't see there being any need for that many decimals... I think 6, maybe 12 is more than enough.

File: aptos-move/framework/aptos-framework/sources/TransactionFee.move (hunk 6 above)
Comment (davidiw) [reply aptos-move/framework/aptos-framework/sources/TransactionFee.move:15]: honestly, we should just put this into account.

File: aptos-move/framework/aptos-framework/sources/TypeInfo.move (hunk 7 above)
Comment (davidiw) [reply aptos-move/framework/aptos-framework/sources/TypeInfo.move:24]: this is unintuitive to me.

File: aptos-move/framework/cached-packages/src/aptos_stdlib.rs [hunk 12]
```diff
@@ -3,3 +3,33 @@
 
 #![allow(unused_imports)]
 include!(concat!(env!("OUT_DIR"), "/transaction_script_builder.rs"));
+
+use move_deps::move_core_types::language_storage::{StructTag, TypeTag};
```
Comment (davidiw) [inline aptos-move/framework/cached-packages/src/aptos_stdlib.rs:7]: clever

File: aptos-move/transaction-builder-generator/tests/generation.rs [hunk 13]
```diff
@@ -231,6 +231,7 @@ test = false
 }
 
 #[test]
+#[ignore]
```
Comment (davidiw) [inline aptos-move/transaction-builder-generator/tests/generation.rs:234]: ?

File: crates/aptos-faucet/src/main.rs (hunk 8 above)
Comment (davidiw) [reply crates/aptos-faucet/src/main.rs:277]: No, it cannot, isn't that why you had to exclude genesis.move @zekun000 ?

File: types/src/account_config/resources/balance.rs [hunk 14]
```diff
@@ -1,9 +1,12 @@
 // Copyright (c) Aptos
 // SPDX-License-Identifier: Apache-2.0
 
+use crate::event::EventHandle;
```
Comment (davidiw) [inline types/src/account_config/resources/balance.rs:4]: we could probably put more global accessors here too

**Review by davidiw — COMMENTED** [review]
this is phenomenal

File: crates/aptos-faucet/src/main.rs (hunk 8 above)
Comment (zekun000) [reply crates/aptos-faucet/src/main.rs:277]: nope, that was because it can’t handle vector<vector<u8>> or vector<u64> etc

File: aptos-move/framework/aptos-framework/sources/TestCoin.move (hunk 11 above)
Comment (borispovod) [reply aptos-move/framework/aptos-framework/sources/TestCoin.move:40]: For `u64` numbers max value is `18446744073709551615`, so it's only `18446744.073` max supply for coin with 12 decimals.
Yet it's good question how much decimals should be enough

File: aptos-move/framework/aptos-framework/sources/TypeInfo.move (hunk 7 above)
Comment (movekevin) [reply aptos-move/framework/aptos-framework/sources/TypeInfo.move:24]: It doesn't make a lot of sense to introduce a dependency just for a unit test. If we want to keep Account, we should separate this test out into a different file.

File: aptos-move/framework/aptos-framework/sources/TestCoin.move (hunk 11 above)
Comment (movekevin) [reply aptos-move/framework/aptos-framework/sources/TestCoin.move:40]: 18 is the standard decimals for most tokens on other chains. This originated from 1 ETH = 10**18 wei. u64 is definitely a bit limiting but it also depends on how much gas costs/gas schedule. On EVM chains, gas is measured in 10**9 (gwei or nAVAX). I can update this to be 6 for now but we need a broader discussion on decimals, gas costs, introducing u256.

File: aptos-move/transaction-builder-generator/tests/generation.rs (hunk 13 above)
Comment (movekevin) [reply aptos-move/transaction-builder-generator/tests/generation.rs:234]: This test unfortunately can't be fixed without the generator supporting generic typings (same exact prob as with aptos_stdlib but with no possible workaround). I can look into that later. Most tests here are already being skipped anyway.

File: api/src/tests/transactions_test.rs (hunk 10 above)
Comment (movekevin) [reply api/src/tests/transactions_test.rs]: Updated.

File: aptos-move/framework/aptos-framework/sources/Genesis.move (hunk 5 above)
Comment (zekun000) [reply aptos-move/framework/aptos-framework/sources/Genesis.move:104]: we already have one governance account, do we want a different one? it's probably better to separate from 0x1 too

File: aptos-move/framework/aptos-framework/sources/Coin.move [hunk 15]
```diff
@@ -189,6 +223,31 @@ module AptosFramework::Coin {
...
+        );
+
+        move_to(account, mint_cap);
+        move_to(account, burn_cap);
+    }
+
+    /// Creates a new Coin with given `CoinType` and returns minting/burning capabilities.
+    /// The given signer also becomes the account hosting the information
+    /// about the coin (name, supply, etc.).
+    public fun initialize_with_capabilities<CoinType>(
```
Comment (zekun000) [inline aptos-move/framework/aptos-framework/sources/Coin.move:244]: not for this PR, but we should think about how to remove Cap

File: types/src/account_config/resources/balance.rs [hunk 16]
```diff
@@ -15,21 +17,39 @@ use serde::{Deserialize, Serialize};
 #[cfg_attr(any(test, feature = "fuzzing"), derive(Arbitrary))]
 pub struct BalanceResource {
     coin: u64,
+    deposit_events: EventHandle,
```
Comment (zekun000) [inline types/src/account_config/resources/balance.rs:20]: better rename to CoinStoreResource

File: aptos-move/framework/aptos-framework/sources/Coin.move [hunk 17]
```diff
@@ -286,19 +372,19 @@ module AptosFramework::Coin {
...
-    /// Withdraw specifed `amount` of coin from the signing account.
+    /// Withdraw specifed `amount` of coin `CoinType` from the signing account.
     public fun withdraw<CoinType>(
         account: &signer,
         amount: u64,
     ): Coin<CoinType> acquires CoinStore {
         let account_addr = Signer::address_of(account);
         assert!(
-            exists<CoinStore<CoinType>>(account_addr),
+            is_registered_for_account<CoinType>(account_addr),
```
Comment (zekun000) [inline aptos-move/framework/aptos-framework/sources/Coin.move:387]: we need to emit an event for every transaction that pays a gas fee here, I was trying to avoid it

File: aptos-move/e2e-testsuite/src/tests/peer_to_peer.rs [hunk 18]
```diff
@@ -179,12 +175,16 @@ fn few_peer_to_peer_with_event() {
 
             // check events
             for event in txn_output.events() {
-                if let Ok(payload) = SentEvent::try_from(event) {
+                if let Ok(payload) = WithdrawEvent::try_from(event) {
+                    if payload.amount() == 0 {
```
[the remark this answers is not in the document]
Comment (movekevin) [reply aptos-move/e2e-testsuite/src/tests/peer_to_peer.rs:179]: This is because we're also generating withdrawal event for gas burns and in these test setups, gas is 0. This test specifically checks for actual transfers.

File: aptos-move/framework/aptos-framework/sources/Genesis.move (hunk 5 above)
Comment (movekevin) [reply aptos-move/framework/aptos-framework/sources/Genesis.move:104]: This is not exactly a governance account in the sense that we don't control the key for it. This is a special account where system capabilities are stored. I think it makes sense to use 0x1 since all of the system modules are deployed there and it makes sense that the capabilities needed for those modules to interact are also encapsulated there.

File: aptos-move/framework/aptos-framework/sources/Coin.move (hunk 15 above)
Comment (movekevin) [reply aptos-move/framework/aptos-framework/sources/Coin.move:244]: Yes, I'll spend more time on the improving the Coin standard next week (in addition to staking)

File: aptos-move/framework/aptos-framework/sources/Coin.move (hunk 17 above)
Comment (movekevin) [reply aptos-move/framework/aptos-framework/sources/Coin.move:387]: Yes this is one aspect I'm thinking through at the moment. Would love to hear your thoughts on it. To summarize the discussion so far:

Gas burns also generate withdrawal events since TransactionFees calls Coin::withdraw, which in my opinion is acceptable and actually could be useful. If we strongly don't want that from an UX perspective, we'd need to either 
(1) add a functionality to burn coins from an account (with burn capability) or
(2) have an option to allow withdrawal without generating events.

I'm leaning more toward (1) if we want to do that since (2) can open up the door for abuse (“secret” withdrawals that don't generate events).

File: aptos-move/framework/aptos-framework/sources/TestCoin.move [hunk 19]
```diff
@@ -47,44 +30,34 @@ module AptosFramework::TestCoin {
...
+        );
 
-    struct ReceivedEvent has drop, store {
-        amount: u64,
-        from: address,
-    }
+        // Mint the core resource account TestCoin for gas so it can execute system transactions.
+        Coin::register_for_account<TestCoin>(core_resource);
+        let coins = Coin::mint_with_capability<TestCoin>(
+            18446744073709551615,
```
Comment (areshand) [inline aptos-move/framework/aptos-framework/sources/TestCoin.move:47]: what happens if we run out of all testcoin?

File: aptos-move/e2e-tests/src/account.rs [hunk 20]
```diff
@@ -596,7 +598,7 @@ impl AccountData {
     /// Creates a writeset that contains the account data and can be patched to the storage
     /// directly.
     pub fn to_writeset(&self) -> WriteSet {
-        let (account_blob, balance_blob, transfer_events) = self.to_value();
+        let (account_blob, balance_blob) = self.to_value();
```
[the remark this answers is not in the document]
Comment (movekevin) [reply aptos-move/e2e-tests/src/account.rs]: Updated

File: aptos-move/framework/aptos-framework/sources/Coin.move [hunk 21]
```diff
@@ -78,22 +98,27 @@ module AptosFramework::Coin {
...
+            is_registered_for_account<CoinType>(owner),
             Errors::not_published(ECOIN_STORE_NOT_PUBLISHED),
         );
         borrow_global<CoinStore<CoinType>>(owner).coin.value
     }
 
-    /// Returns `true` if the type `CoinType` is a registered coin.
-    public fun is_registered<CoinType>(): bool {
+    /// Returns `true` if the type `CoinType` is an initialized coin.
+    public fun is_initialized<CoinType>(): bool {
```
[the remark this answers is not in the document]
Comment (movekevin) [reply aptos-move/framework/aptos-framework/sources/Coin.move]: Updated

File: types/src/account_config/resources/balance.rs (hunk 16 above)
Comment (movekevin) [reply types/src/account_config/resources/balance.rs:20]: Updated

File: aptos-move/e2e-tests/src/account.rs [hunk 22]
```diff
@@ -317,12 +313,18 @@ impl TransactionBuilder {
 #[derive(Clone, Debug, Eq, PartialEq)]
 pub struct Balance {
```
[the remark this answers is not in the document]
Comment (movekevin) [reply aptos-move/e2e-tests/src/account.rs]: Updated

File: aptos-move/e2e-tests/src/executor.rs [hunk 23]
```diff
@@ -273,7 +269,7 @@ impl FakeExecutor {
 
     /// Reads the balance resource value for an account from this executor's data store with the
     /// given balance currency_code.
-    pub fn read_balance_resource(&self, account: &Account) -> Option<BalanceResource> {
+    pub fn read_balance_resource(&self, account: &Account) -> Option<CoinStoreResource> {
```
[the remark this answers is not in the document]
Comment (movekevin) [reply aptos-move/e2e-tests/src/executor.rs]: Updated

File: aptos-move/framework/aptos-framework/sources/TestCoin.move (hunk 19 above)
Comment (movekevin) [reply aptos-move/framework/aptos-framework/sources/TestCoin.move:47]: In testnet, we do have the key for the core resources account, which has mint capability so we can mint more to it. With that said, there might be a way to avoid having to make transactions from that account, but I'll need to do separate research on that.

File: aptos-move/framework/aptos-framework/sources/Coin.move (hunk 17 above)
Comment (movekevin) [reply aptos-move/framework/aptos-framework/sources/Coin.move]: I updated with approach (1), adding a burn_from_account function to Coin.move that allows someone with BurnCapability to burn coins directly from an account. We can discuss later on whether it makes sense for this functionality to be in the standard Coin vs being in a different extended Coin version (e.g. BurnableCoin)

File: aptos-move/e2e-testsuite/src/tests/peer_to_peer.rs (hunk 18 above)
[the remark this answers is not in the document]
Comment (movekevin) [reply aptos-move/e2e-testsuite/src/tests/peer_to_peer.rs]: Updated now that gas burns no longer emit withdrawal events.

**Comment by movekevin** [comment]
@davidiw  Added python/rust code (typescript already added in a previous PR) + tutorial page for the first_coin example.

File: aptos-move/e2e-tests/src/account.rs [hunk 24]
```diff
@@ -134,11 +134,7 @@ impl Account {
     ///
     /// Use this to retrieve or publish the Account balance blob.
     pub fn make_balance_access_path(&self) -> AccessPath {
```
Comment (davidiw) [inline aptos-move/e2e-tests/src/account.rs:136]: rename to make_coin_store_access_path

File: aptos-move/framework/aptos-framework/sources/Coin.move (hunk 15 above)
Comment (davidiw) [reply aptos-move/framework/aptos-framework/sources/Coin.move:256]: I don't follow?

File: aptos-move/framework/aptos-framework/sources/Genesis.move (hunk 5 above)
Comment (davidiw) [reply aptos-move/framework/aptos-framework/sources/Genesis.move:104]: I don't know if account separation matters so much, but it probably doesn't hurt to separate shared governance accounts based upon the role...

* Framework
* Staking / rewards
* Upgrades / versions

but that feels like a conversation that requires more thought

File: aptos-move/framework/aptos-framework/sources/TestCoin.move (hunk 19 above)
Comment (davidiw) [reply aptos-move/framework/aptos-framework/sources/TestCoin.move:47]: this account is only paying gas to create new accounts

File: aptos-move/framework/aptos-framework/sources/TypeInfo.move (hunk 7 above)
Comment (davidiw) [reply aptos-move/framework/aptos-framework/sources/TypeInfo.move:24]: I imagine this dependency isn't even being compiled in. Lots of tests pull in dependencies just for testing.

File: aptos-move/transaction-builder-generator/tests/generation.rs (hunk 13 above)
Comment (davidiw) [reply aptos-move/transaction-builder-generator/tests/generation.rs:234]: leave a comment then

File: developer-docs-site/static/examples/python/first_coin.py [hunk 25]
```diff
@@ -0,0 +1,128 @@
...
+        """Publish a new module to the blockchain within the specified account"""
+
+        payload = {
+            "type": "module_bundle_payload",
+            "modules": [
+                {"bytecode": f"0x{module_hex}"},
+            ],
+        }
+        res = self.execute_transaction_with_payload(account_from, payload)
+        return str(res["hash"])
```
Comment (davidiw) [inline developer-docs-site/static/examples/python/first_coin.py:23]: why not import first move module? also should look at cli

File: types/src/account_config/resources/balance.rs [hunk 26]
```diff
@@ -1,9 +1,11 @@
 // Copyright (c) Aptos
```
Comment (davidiw) [inline types/src/account_config/resources/balance.rs:1]: file name needs to be updated

**Review by davidiw — APPROVED** [review]
wow, lots of nits, but overall fantastic!!! unblocking -- but if you need another review let me know.

note: we landed @borispovod PR on Coin fixes.

File: aptos-move/framework/aptos-framework/sources/TypeInfo.move (hunk 7 above)
Comment (movekevin) [reply aptos-move/framework/aptos-framework/sources/TypeInfo.move:24]: Compiler complained about cyclic dependency before, which was the reason why I had to change this.

File: aptos-move/framework/aptos-framework/sources/Coin.move (hunk 15 above)
Comment (movekevin) [reply aptos-move/framework/aptos-framework/sources/Coin.move:256]: Zekun is talking about adding functionalities to remove/destroy capabilities, which I think can be useful in 2 cases:

1. Capabilities can currently be copied and handed to different accounts. There might be a need to remove these.
2. There can be a Coin that has fixed supply and not mintable/burnable. So the capabilities created would need to be destroyed.

File: aptos-move/framework/aptos-framework/sources/TestCoin.move (hunk 19 above)
Comment (movekevin) [reply aptos-move/framework/aptos-framework/sources/TestCoin.move:47]: I'm not sure if there's anything actionable for now. We can discuss/look into potentially removing this or at least minting less separately.

File: types/src/account_config/resources/balance.rs (hunk 26 above)
Comment (movekevin) [reply types/src/account_config/resources/balance.rs]: Updated

File: aptos-move/e2e-tests/src/account.rs (hunk 24 above)
Comment (movekevin) [reply aptos-move/e2e-tests/src/account.rs]: Updated
